### PR TITLE
Add a configurable timeout to the sync process

### PR DESCRIPTION
**Pull request by george-angel**
Fixes https://github.com/kubernetes/git-sync/issues/115

### CODE REVIEW

File: cmd/git-sync/main.go [hunk 1]
```diff
@@ -181,7 +184,9 @@ func main() {
 	initialSync := true
 	failCount := 0
 	for {
-		if err := syncRepo(*flRepo, *flBranch, *flRev, *flDepth, *flRoot, *flDest); err != nil {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*time.Duration(*flSyncTimeout))
+		err := syncRepo(ctx, *flRepo, *flBranch, *flRev, *flDepth, *flRoot, *flDest)
```
Comment (thockin) [inline cmd/git-sync/main.go]: Why did you change the style from `if err:= ...; err != nil` ?  I don't see a need for err beyond this block?

Comment (george-angel) [reply cmd/git-sync/main.go]: Accidental, reverted.

**Comment by thockin** [comment]
Thanks.  Code LGTM.  Only tests remain.

**Comment by thockin** [comment]
Thanks for poking at it.   Hypothetical:

Add a `--git` flag which is the path to a git binary.  Use that instead of "git".

Add a shell script that is:

```
#!/bin/sh
sleep 1
git "$@"
```

In the e2e test, write a case that uses that flag and that script to sync, with a timeout that is too low.  Verify failure.

Would it help if I added the flag myself?  It's possible I am missing something that makes this more complicated.

**Comment by george-angel** [comment]
Nope, I was unsure what you meant by `--git` flag, that makes sense.

**Comment by thockin** [comment]
I just whipped up a change to add that flag, it's pretty straight forward, it seems.

**Comment by thockin** [comment]
```
diff --git a/cmd/git-sync/main.go b/cmd/git-sync/main.go
index 87bd21f..1cd8295 100644
--- a/cmd/git-sync/main.go
+++ b/cmd/git-sync/main.go
@@ -72,6 +72,9 @@ var flSSHKnownHosts = flag.Bool("ssh-known-hosts", envBool("GIT_KNOWN_HOSTS", tr
 var flCookieFile = flag.Bool("cookie-file", envBool("GIT_COOKIE_FILE", false),
        "use git cookiefile")
 
+var flGitCmd = flag.String("git", envString("GIT_SYNC_GIT", "git"),
+       "the git command to run (subject to PATH search)")
+
 var log = newLoggerOrDie()
 
 func newLoggerOrDie() logr.Logger {
@@ -144,8 +147,8 @@ func main() {
                flag.Usage()
                os.Exit(1)
        }
-       if _, err := exec.LookPath("git"); err != nil {
-               fmt.Fprintf(os.Stderr, "ERROR: git executable not found: %v\n", err)
+       if _, err := exec.LookPath(*flGitCmd); err != nil {
+               fmt.Fprintf(os.Stderr, "ERROR: git executable %q not found: %v\n", *flGitCmd, err)
                os.Exit(1)
        }
 
@@ -269,7 +272,7 @@ func updateSymlink(gitRoot, link, newDir string) error {
 
                log.V(1).Infof("removed %s", currentDir)
 
-               _, err := runCommand(gitRoot, "git", "worktree", "prune")
+               _, err := runCommand(gitRoot, *flGitCmd, "worktree", "prune")
                if err != nil {
                        return err
                }
@@ -285,13 +288,13 @@ func addWorktreeAndSwap(gitRoot, dest, branch, rev, hash string) error {
        log.V(0).Infof("syncing to %s (%s)", rev, hash)
 
        // Update from the remote.
-       if _, err := runCommand(gitRoot, "git", "fetch", "--tags", "origin", branch); err != nil {
+       if _, err := runCommand(gitRoot, *flGitCmd, "fetch", "--tags", "origin", branch); err != nil {
                return err
        }
 
        // Make a worktree for this exact git hash.
        worktreePath := path.Join(gitRoot, "rev-"+hash)
-       _, err := runCommand(gitRoot, "git", "worktree", "add", worktreePath, "origin/"+branch)
+       _, err := runCommand(gitRoot, *flGitCmd, "worktree", "add", worktreePath, "origin/"+branch)
        if err != nil {
                return err
        }
@@ -311,7 +314,7 @@ func addWorktreeAndSwap(gitRoot, dest, branch, rev, hash string) error {
        }
 
        // Reset the worktree's working copy to the specific rev.
-       _, err = runCommand(worktreePath, "git", "reset", "--hard", hash)
+       _, err = runCommand(worktreePath, *flGitCmd, "reset", "--hard", hash)
        if err != nil {
                return err
        }
@@ -334,7 +337,7 @@ func cloneRepo(repo, branch, rev string, depth int, gitRoot string) error {
                args = append(args, "--depth", strconv.Itoa(depth))
        }
        args = append(args, repo, gitRoot)
-       _, err := runCommand("", "git", args...)
+       _, err := runCommand("", *flGitCmd, args...)
        if err != nil {
                if strings.Contains(err.Error(), "already exists and is not an empty directory") {
                        // Maybe a previous run crashed?  Git won't use this dir.
@@ -343,7 +346,7 @@ func cloneRepo(repo, branch, rev string, depth int, gitRoot string) error {
                        if err != nil {
                                return err
                        }
-                       _, err = runCommand("", "git", args...)
+                       _, err = runCommand("", *flGitCmd, args...)
                        if err != nil {
                                return err
                        }
@@ -357,7 +360,7 @@ func cloneRepo(repo, branch, rev string, depth int, gitRoot string) error {
 }
 
 func hashForRev(rev, gitRoot string) (string, error) {
-       output, err := runCommand(gitRoot, "git", "rev-parse", rev)
+       output, err := runCommand(gitRoot, *flGitCmd, "rev-parse", rev)
        if err != nil {
                return "", err
        }
@@ -439,7 +442,7 @@ func getRevs(localDir, branch, rev string) (string, string, error) {
 }
 
 func remoteHashForRef(ref, gitRoot string) (string, error) {
-       output, err := runCommand(gitRoot, "git", "ls-remote", "-q", "origin", ref)
+       output, err := runCommand(gitRoot, *flGitCmd, "ls-remote", "-q", "origin", ref)
        if err != nil {
                return "", err
        }
@@ -476,13 +479,13 @@ func runCommand(cwd, command string, args ...string) (string, error) {
 
 func setupGitAuth(username, password, gitURL string) error {
        log.V(1).Infof("setting up the git credential cache")
-       cmd := exec.Command("git", "config", "--global", "credential.helper", "cache")
+       cmd := exec.Command(*flGitCmd, "config", "--global", "credential.helper", "cache")
        output, err := cmd.CombinedOutput()
        if err != nil {
                return fmt.Errorf("error setting up git credentials %v: %s", err, string(output))
        }
 
-       cmd = exec.Command("git", "credential", "approve")
+       cmd = exec.Command(*flGitCmd, "credential", "approve")
        stdin, err := cmd.StdinPipe()
        if err != nil {
                return err
@@ -542,7 +545,7 @@ func setupGitCookieFile() error {
                return fmt.Errorf("error: could not find git cookie file: %v", err)
        }
 
-       cmd := exec.Command("git", "config", "--global", "http.cookiefile", pathToCookieFile)
+       cmd := exec.Command(*flGitCmd, "config", "--global", "http.cookiefile", pathToCookieFile)
        output, err := cmd.CombinedOutput()
        if err != nil {
                return fmt.Errorf("error configuring git cookie file %v: %s", err, string(output))
```

**Comment by george-angel** [comment]
Thanks, however `make test` is failing for me on the current master:

```
Unable to find image 'e2e/git-sync:v3.0.1-6-g205c771-dirty' locally
docker: Error response from daemon: pull access denied for e2e/git-sync, repository does not exist or may require 'docker login'.
See 'docker run --help'.
```

**Comment by thockin** [comment]
Ugh, I missed something when adding manifest support.

I'll post a patch here momentarily that will unblock you, and will merge
that upstream

**Comment by thockin** [comment]
```
diff --git a/test_e2e.sh b/test_e2e.sh
index 2d25300..b281546 100755
--- a/test_e2e.sh
+++ b/test_e2e.sh
@@ -83,7 +83,7 @@ function GIT_SYNC() {
         -u $(id -u):$(id -g) \
         -v "$DIR":"$DIR" \
         --rm \
-        e2e/git-sync:$(make -s version) \
+        e2e/git-sync:$(make -s version)__$(go env GOOS)_$(go env GOARCH) \
         "$@"
 }
```

**Comment by thockin** [comment]
merged upstream, you can just rebase

**Comment by george-angel** [comment]
@thockin last commit includes what I think are relevant tests. Thanks.
